tools/harness-simulation/posix/harness_dev_discovery.py

```diff
@@ -0,0 +1,169 @@
+    addrinfo = socket.getaddrinfo(group, None)[0]
+    assert addrinfo[0] == socket.AF_INET6
+
+    # Create a socket
+    s = socket.socket(addrinfo[0], socket.SOCK_DGRAM)
+    s.setsockopt(socket.SOL_SOCKET, socket.SO_BINDTODEVICE, (ifname + '\0').encode('ascii'))
+
+    # Bind it to the port
+    s.bind(('', port))
+
+    group_bin = socket.inet_pton(addrinfo[0], addrinfo[4][0])
+    # Join group
+    interface_index = if_nametoindex(ifname)
+    mreq = group_bin + struct.pack('@I', interface_index)
+    s.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_JOIN_GROUP, mreq)
+
+    return s
+
+
+def advertise_ftd(s: socket.socket, dst, ven: str, ver: str, add: str, por: int, number: int):
+    # Node ID of ot-cli-ftd is 1-indexed
+    for i in range(1, number + 1):
+        info = {
+            'ven': ven,
+            'mod': f'{ven}_{i}',
+            'ver': ver,
+            'add': f'{i}@{add}',
+            'por': por,
+        }
+        logging.info('Advertise: %r', info)
+        s.sendto(json.dumps(info).encode('utf-8'), dst)
+
+
+def advertise_sniffer(s: socket.socket, dst, add: str, number: int):
+    for i in range(number):
+        info = 's%d@%s' % (i, add)
+        logging.info('Advertise: %r', info)
+        s.sendto(info.encode('utf-8'), dst)
+
+
+def main():
+    logging.basicConfig(level=logging.INFO)
+
+    # Parse arguments
+    parser = argparse.ArgumentParser()
+
+    # Determine the interface
+    parser.add_argument('-i',
+                        '--interface',
+                        dest='ifname',
+                        type=str,
+                        required=True,
+                        help='the interface used for discovery')
+
+    # Determine the number of OpenThread 1.1 FTD simulations to be "detected" and then initiated
+    parser.add_argument('--ot1.1',
+                        dest='ot11_num',
+                        type=int,
+                        required=False,
+                        default=0,
+                        help=f'the number of OpenThread FTD simulations, no more than {MAX_OT11_NUM}')
+
+    # Determine the number of sniffer simulations to be initiated and then detected
+    parser.add_argument('-s',
+                        '--sniffer',
+                        dest='sniffer_num',
+                        type=int,
+                        required=False,
+                        default=0,
+                        help=f'the number of sniffer simulations, no more than {MAX_SNIFFER_NUM}')
+
+    args = parser.parse_args()
+
+    # Check validation of arguments
+    if not (0 <= args.ot11_num and args.ot11_num <= MAX_OT11_NUM):
+        raise ValueError(f'The number of FTDs should be between 0 and {MAX_OT11_NUM}')
+
+    if not (0 <= args.sniffer_num and args.sniffer_num <= MAX_SNIFFER_NUM):
+        raise ValueError(f'The number of FTDs should be between 0 and {MAX_SNIFFER_NUM}')
+
+    if args.ot11_num == 0 and args.sniffer_num == 0:
```

Comment:
We also need to check that ot11_num <= 33?

Reply:
It was checked in the line 136-137.